xdsl/transforms/experimental/stencil_global_to_local.py

```diff
@@ -0,0 +1,491 @@
+from dataclasses import dataclass
+from typing import TypeVar, Iterable, ClassVar
+from abc import ABC, abstractmethod
+from math import prod
+
+from xdsl.utils.hints import isa
+from xdsl.pattern_rewriter import (PatternRewriter, PatternRewriteWalker,
+                                   RewritePattern, GreedyRewritePatternApplier,
+                                   op_type_rewrite_pattern)
+from xdsl.ir import MLContext, Operation, SSAValue, Block, Region
+from xdsl.irdl import Attribute
+from xdsl.dialects import builtin, mpi, memref, arith, scf
+from xdsl.dialects.experimental import stencil
+
+_T = TypeVar('_T', bound=Attribute)
+
+
+@dataclass
+class HaloExchangeDef:
+    """
+    This declares a region to be "halo-exchanged".
+    The semantics define that the region specified by offset and size
+    is the *received part*. To get the section that should be sent,
+    use the source_area() method to get the source area.
+
+    offset gives the coordinates from the origin of the stencil field.
+
+    size gives the size of the buffer to be exchanged.
+
+    source_offset gives a translation (n-d offset) where the data should be
+    read from that is exchanged with the other node.
+
+    Finally, neighbor gives the n-dimensional offset to the node with whom
+    this edge should be exchanged.
+
+    Example:
+
+        offset = [4, 0]
+        size   = [10, 1]
+        source_offset = [0, 1]
+        neighbor = [-1, 0]
+
+    To visualize:
+    a0  b0        c0
+        xxxxxxxxxx    a1
+        oooooooooo    b1
+
```

Comment:
This looks like written a bit in a hurry, maybe adding a TODO to improve at some point, would be good